decoder/vaapidecoder_h265.cpp

```diff
@@ -0,0 +1,1784 @@
+        num_lt_pics = slice_hdr->num_long_term_sps + slice_hdr->num_long_term_pics;
+        /* The variables PocLsbLt[i] and UsedByCurrPicLt[i] are derived as follows: */
+        for (i = 0; i < num_lt_pics; i++) {
+            if (i < slice_hdr->num_long_term_sps) {
+                PocLsbLt[i] = sps->lt_ref_pic_poc_lsb_sps[slice_hdr->lt_idx_sps[i]];
+                UsedByCurrPicLt[i] =
+                sps->used_by_curr_pic_lt_sps_flag[slice_hdr->lt_idx_sps[i]];
+            } else {
+                PocLsbLt[i] = slice_hdr->poc_lsb_lt[i];
+                UsedByCurrPicLt[i] = slice_hdr->used_by_curr_pic_lt_flag[i];
+            }
+            if (UsedByCurrPicLt[i])
+                numtotalcurr++;
+        }
+
+        m_NumPocTotalCurr = numtotalcurr;
+
+        /* The variable DeltaPocMsbCycleLt[i] is derived as follows: (7-38) */
+        for (i = 0; i < num_lt_pics; i++) {
+            if (i == 0 || i == slice_hdr->num_long_term_sps)
+                DeltaPocMsbCycleLt[i] = slice_hdr->delta_poc_msb_cycle_lt[i];
+            else
+                DeltaPocMsbCycleLt[i] =
+                slice_hdr->delta_poc_msb_cycle_lt[i] + DeltaPocMsbCycleLt[i - 1];
+        }
+
+        /* (8-5) */
+        for (i = 0, j = 0, k = 0; i < num_lt_pics; i++) {
+            pocLt = PocLsbLt[i];
+            if (slice_hdr->delta_poc_msb_present_flag[i])
+                pocLt +=
+                picture->m_poc - DeltaPocMsbCycleLt[i] * MaxPicOrderCntLsb -
+                slice_hdr->pic_order_cnt_lsb;
+            if (UsedByCurrPicLt[i]) {
+                m_PocLtCurr[j] = pocLt;
+                CurrDeltaPocMsbPresentFlag[j++] =
+                slice_hdr->delta_poc_msb_present_flag[i];
+            } else {
+                m_PocLtFoll[k] = pocLt;
+                FollDeltaPocMsbPresentFlag[k++] =
+                slice_hdr->delta_poc_msb_present_flag[i];
+            }
+        }
+        m_NumPocLtCurr = j;
+        m_NumPocLtFoll = k;
+
+    }
+
+    /* the derivation process for the RPS and the picture marking */
+    derive_and_mark_rps (picture.get(), CurrDeltaPocMsbPresentFlag, FollDeltaPocMsbPresentFlag);
+
+    return true;
+}
+
+void VaapiDecoderH265::init_picture_refs(VaapiDecPictureH265 * picture, H265SliceHdr * slice_hdr)
+{
+    uint32_t NumRpsCurrTempList0 = 0, NumRpsCurrTempList1 = 0;
+    VaapiDecPictureH265 *RefPicListTemp0[16] = { NULL, }, *RefPicListTemp1[16] = {
+    NULL,};
+    uint32_t i, rIdx = 0;
+    uint32_t num_ref_idx_l0_active_minus1 = 0;
+    uint32_t num_ref_idx_l1_active_minus1 = 0;
+    H265RefPicListModification *ref_pic_list_modification;
+    uint32_t type;
+
+    memset (m_RefPicList0, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    memset (m_RefPicList1, 0, sizeof (VaapiDecPictureH265 *) * 16);
+    m_RefPicList0_count = m_RefPicList1_count = 0;
+
+    if ((slice_hdr->type != H265_B_SLICE) &&
+        (slice_hdr->type != H265_P_SLICE))
+        return;
+
+    if (slice_hdr->dependent_slice_segment_flag) {
```

Comment:
better if we can use this

H265SliceHdr* hdr;
if (slice_hdr->dependent_slice_segment_flag)
    hdr = m_prev_independent_slice;
else
   hdr = slice_hdr;
num_ref_idx_l0_active_minus1 = hdr->num_ref_idx_l0_active_minus1;
.....